abci/app.go

```diff
@@ -0,0 +1,377 @@
+		if err == nil {
+			proposedTxs = append(proposedTxs, tx)
+		}
+		// Adding the curse words from vote extensions too
+		if !IsCurseWord(msg.Message, voteExtensionCurseWords) {
+			proposedTxs = append(proposedTxs, tx)
+		} else {
+			banTx := model.BanTx{UserName: msg.Sender}
+			bannedUsersString[msg.Sender] = struct{}{}
+			resultBytes, err := json.Marshal(banTx)
+			if err == nil {
+				finalProposal = append(finalProposal, resultBytes)
+			} else {
+				panic(fmt.Errorf("invalid ban transaction in prepareProposal"))
+			}
+		}
+	}
+
+	// Need to loop again through the proposed Txs to make sure there is none left by a user that was banned after the tx was accepted
+	for _, tx := range proposedTxs {
+		// there should be no error here as these are just transactions we have checked and added
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			panic(err)
+		}
+		if _, ok := bannedUsersString[msg.Sender]; !ok {
+			finalProposal = append(finalProposal, tx)
+		}
+	}
+	return &types.ResponsePrepareProposal{Txs: finalProposal}, nil
+}
+
+func (ForumApp) ProcessProposal(_ context.Context, processproposal *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+	fmt.Println("entered processProp")
+	bannedUsers := make(map[string]struct{}, 0)
+
+	finishedBanTxIdx := len(processproposal.Txs)
+	for i, tx := range processproposal.Txs {
+		if isBanTx(tx) {
+			var parsedBan model.BanTx
+			err := json.Unmarshal(tx, &parsedBan)
+			if err != nil {
+				return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+			}
+			bannedUsers[parsedBan.UserName] = struct{}{}
+		} else {
+			finishedBanTxIdx = i
+			break
+		}
+	}
+
+	for _, tx := range processproposal.Txs[finishedBanTxIdx:] {
+		msg, err := model.ParseMessage(tx)
+		if err != nil {
+			return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+		}
+		if _, ok := bannedUsers[msg.Sender]; ok {
+			// sending us a tx from a banned user
+			return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_REJECT}, nil
+		}
+	}
+	return &types.ResponseProcessProposal{Status: types.ResponseProcessProposal_ACCEPT}, nil
+}
+
+// Deliver the decided block with its txs to the Application
+func (app *ForumApp) FinalizeBlock(_ context.Context, req *abci.RequestFinalizeBlock) (*abci.ResponseFinalizeBlock, error) {
+	fmt.Println("entered finalizeBlock")
+	// Iterate over Tx in current block
+	app.onGoingBlock = app.state.DB.GetDB().NewTransaction(true)
+	respTxs := make([]*types.ExecTxResult, len(req.Txs))
+	finishedBanTxIdx := len(req.Txs)
+	for i, tx := range req.Txs {
+		var err error
+
+		if isBanTx(tx) {
+			banTx := new(model.BanTx)
+			err = json.Unmarshal(tx, &banTx)
+			if err != nil {
+				respTxs[i] = &types.ExecTxResult{Code: CodeTypeEncodingError}
+			} else {
+				err := UpdateOrSetUser(app.state.DB, banTx.UserName, true, app.onGoingBlock)
+				if err != nil {
+					panic(err)
```

Comment:
are we sure we want to panic here, rather than report an error in `ExecTxResult`?